include/votca/xtp/bseoperator_btda.h

```diff
@@ -0,0 +1,232 @@
+  }
+
+  // this is not a fast method
+  const double& operator()(const size_t i, const size_t j) const {
+    if (i==j) {
+      return _diag(i);
+    } else {
+      Eigen::RowVectorXd row_out = row(i);
+      return  row_out(j);
+    }
+  };
+
+  //  get a row of the operator
+  Eigen::RowVectorXd row(int index) const {
+    /* Returna row of the operator 
+    H = [ A   B
+         -B* -A* ]
+    */
+    int lsize = this->_size;
+    int halfsize = lsize/2;
+    Eigen::RowVectorXd row_out = Eigen::RowVectorXd::Zero(lsize);
+
+    if (index < halfsize)
+    {
+
+      Eigen::RowVectorXd a = this->_A.row(index);
+      Eigen::RowVectorXd b = this->_B.row(index);    
+      row_out.head(halfsize) = a;
+      row_out.tail(halfsize) = b;
+
+    } else {
+
+      Eigen::RowVectorXd a = this->_A.row(index-halfsize);
+      Eigen::RowVectorXd b = this->_B.row(index-halfsize);    
+      row_out.head(halfsize) = -b;
+      row_out.tail(halfsize) = -a;
+    }
+
+    return row_out;
+  }
+
+  Eigen::VectorXd get_diagonal() const {
+    Eigen::VectorXd diag = Eigen::VectorXd::Zero(_size);
+#pragma omp parallel for
+    for (int i=0; i < _size; i++) {
+      Eigen::RowVectorXd r = row(i);
+      diag(i) = r(i);
+    }
+    return diag;
+  }
+
+  Eigen::VectorXd diagonal() const {return _diag;}
+  
+  // get the full matrix if we have to
+  Eigen::MatrixXd get_full_matrix() const {
+    Eigen::MatrixXd matrix = Eigen::MatrixXd::Zero(_size, _size);
+  
+    for (int i = 0; i < _size; i++) {
+      matrix.row(i) = row(i);
+    }
+    return matrix;
+  }
+
+ private:
+
+  const MatrixReplacementA & _A;
+  const MatrixReplacementB & _B;
+  int _size;
+  Eigen::VectorXd _diag; 
+
+};
+}  // namespace xtp
+}  // namespace votca
+
+namespace Eigen {
+namespace internal {
+
+// replacement of the mat*vect operation
+template <typename Vtype, typename MatrixReplacementA, typename MatrixReplacementB>
+struct generic_product_impl<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Vtype, DenseShape,
+                            DenseShape, GemvProduct>
+    : generic_product_impl_base<
+          votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Vtype,
+          generic_product_impl<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Vtype> >{
+
+  typedef
+      typename Product<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Vtype>::Scalar Scalar;
+
+  template <typename Dest>
+  static void scaleAndAddTo(Dest& dst, const votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>& op,
+                            const Vtype& v, const Scalar& alpha) {
+    // returns dst = alpha * op * v
+    // alpha must be 1 here
+    assert(alpha == Scalar(1) && "scaling is not implemented");
+    EIGEN_ONLY_USED_FOR_DEBUG(alpha);
+
+// make the mat vect product
+#pragma omp parallel for
+    for (int i = 0; i < op.rows(); i++) {
+      dst(i) = op.row(i) * v;
```

Comment:
I think we should leave this to the MatrixReplacemeent Stuff.. i.e just use the blocks indepedently

Reply:
i.e.  result.head()=A*v.head()+B*v.tail() 
result.tail()=-B*v.head()-A*v.tail()

Reply:
Why ? I find the current implementation more clear. and it's the same as for the other matrix replacement ...

Reply:
we would remove .row() entirely ...

Reply:
done